relay/kusama/src/impls.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use core::marker::PhantomData;
+use frame_support::traits::{tokens::ConversionFromAssetBalance, Contains};
+use primitives::Id as ParaId;
+use xcm_builder::IsChildSystemParachain;
+
+/// Determines if the given `asset_kind` is a native asset. If it is, returns the balance without
+/// conversion; otherwise, delegates to the implementation specified by `I`.
+pub struct NativeOnSystemParachain<I>(PhantomData<I>);
+impl<I> ConversionFromAssetBalance<Balance, VersionedLocatableAsset, Balance>
+	for NativeOnSystemParachain<I>
+where
+	I: ConversionFromAssetBalance<Balance, VersionedLocatableAsset, Balance>,
+{
+	type Error = ();
+	fn from_asset_balance(
+		balance: Balance,
+		asset_kind: VersionedLocatableAsset,
+	) -> Result<Balance, Self::Error> {
+		use VersionedLocatableAsset::*;
+		let (location, asset_id) = match asset_kind.clone() {
+			V3 { location, asset_id } => (location.try_into()?, asset_id.try_into()?),
+			V4 { location, asset_id } => (location, asset_id),
+		};
+		if asset_id.0.contains_parents_only(1) &&
+			IsChildSystemParachain::<ParaId>::contains(&location)
+		{
+			Ok(balance)
+		} else {
+			I::from_asset_balance(balance, asset_kind).map_err(|_| ())
```

Comment:
This hinders debuggability. Can you please log the error?

Reply:
In this particular case, upstream we do not really care about type of error and there is only one reason for error.
But for a general implementation I will add the log.